src/sync/incomingSyncUtils.js

```diff
@@ -12,117 +12,244 @@ import {
-  const syncType = syncRecord.SyncType;
-  const recordType = syncRecord.RecordType;
+export function integrateobject(database, settings, syncobject) {
+  // If the sync object is missing either data, object type, sync type, or object ID, ignore
+  if (!syncobject.objectType || !syncobject.SyncType) return;
+  const syncType = syncobject.SyncType;
+  const objectType = syncobject.objectType;
   const changeType = SYNC_TYPES.translate(syncType, EXTERNAL_TO_INTERNAL);
-  const internalRecordType = RECORD_TYPES.translate(recordType, EXTERNAL_TO_INTERNAL);
+  const internalobjectType = object_TYPES.translate(objectType, EXTERNAL_TO_INTERNAL);
+  if (changeType === 'merge') {
+    mergeobjects(database, settings, syncobject);
+  }
 
   switch (changeType) {
     case CHANGE_TYPES.CREATE:
     case CHANGE_TYPES.UPDATE:
-      if (!syncRecord.data) return; // If missing data representing record, ignore
-      createOrUpdateRecord(database, settings, internalRecordType, syncRecord.data);
+      if (!syncobject.data) return; // If missing data representing object, ignore
+      createOrUpdateobject(database, settings, internalobjectType, syncobject.data);
       break;
     case CHANGE_TYPES.DELETE:
-      if (!syncRecord.RecordID) return; // If missing record id, ignore
-      deleteRecord(database, internalRecordType, syncRecord.RecordID);
+      if (!syncobject.objectID) return; // If missing object id, ignore
+      deleteobject(database, internalobjectType, syncobject.objectID);
       break;
     default:
-      return; // Silently ignore change types we don't handle, e.g. merges
+      return;
+  }
+}
+
+/**
+ * Merge two existing objects. Two-level lookup table for 1) the objects related to the type
+ * of the merging objects 2) the fields of the related object to update. Point related objects
+ * of the merged object to point to the kept object and delete.
+ * @param {Realm}  database    The local database
+ * @param {object} settings    Access to app settings
+ * @param {object} mergeobject Data representing the sync object
+ */
+export function mergeobjects(database, settings, mergeobject) {
+  const object_TYPE_TO_TABLES = {
+    item: {
+      StocktakeItem: 'item',
+      TransactionItem: 'item',
+      ItemBatch: 'item',
+      RequisitionItem: 'item',
+    },
+    name: {
+      ItemBatch: 'supplier',
+      Transaction: 'otherParty',
+      Requisition: 'otherStoreName',
+    },
+  };
+
+  const mergedObjectsInternalType = object_TYPES.translate(
+    mergeobject.objectType,
+    EXTERNAL_TO_INTERNAL,
+  );
+
+  const objectToKeep = database
+    .objects(mergedObjectsInternalType)
+    .filtered('id == $0', mergeobject.mergeIDtokeep)[0];
+  const objectToMerge = database
+    .objects(mergedObjectsInternalType)
+    .filtered('id == $0', mergeobject.mergeIDtodelete)[0];
+
+  if (!(objectToKeep && objectToMerge)) {
+    return;
+  }
+
+  const tablesToUpdate = object_TYPE_TO_TABLES[mergeobject.objectType];
+
+  if (tablesToUpdate) {
```

Comment:
To avoid further indentations, i would return from method if !tablesToUpdate. Can also add a todo comment if !tablesToUpdate (`// TODO log to bugsnag if merging not implemented for a certain recordType`)

Reply:
Bugsnag TODO comment added when no tablesToUpdate.